File: LeetCode/Patterns/Backtracking/39. Combination Sum/Solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        ans = []
        path = []
        def memo(i : int, target: int)->None:
            if target == 0:
                ans.append(path.copy())
                return
            if target < 0 or i >= n:
                return

            path.append(candidates[i])
            memo(i, target-candidates[i])
            path.pop()

            memo(i+1, target)
        
        memo(0, target)
        return ans

File: LeetCode/Patterns/Backtracking/39. Combination Sum/Solution.cpp
class Solution {
public:

    void makeCombination(vector<int>& arr, vector<int>& curr, vector<vector<int>>& res, int remSum, int index) {
        if(remSum == 0) {
            res.push_back(curr);
            return;
        }
        if(remSum < 0 || index >= arr.size()) 
            return;

        curr.push_back(arr[index]);
        makeCombination(arr, curr, res, remSum-arr[index], index);
        curr.pop_back();
        makeCombination(arr, curr, res, remSum, index+1);
        
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> curr;
        vector<vector<int>> res;

        makeCombination(candidates, curr, res, target, 0);
        return res;
    }
};